File: src/main/java/sharev/gathering/entity/IntroduceTemplateContent.java
package sharev.gathering.entity;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import sharev.gathering.exception.TemplateVariableMismatchException;

public record IntroduceTemplateContent(
        @NotBlank
        String text,

        @NotNull
        Map<String, String> fieldPlaceholders) {

    private static final Pattern VARIABLE_PATTERN = Pattern.compile("\\$\\{([^}]+)}");

    public IntroduceTemplateContent {
        validateContentAndFieldPlaceholders(text, fieldPlaceholders);
    }

    private static void validateContentAndFieldPlaceholders(String text, Map<String, String> fieldPlaceholders) {
        Matcher matcher = VARIABLE_PATTERN.matcher(text);
        Set<String> foundKeys = new HashSet<>();

        while (matcher.find()) {
            String key = matcher.group(1);
            foundKeys.add(key);

            if (!fieldPlaceholders.containsKey(key)) {
                throw new TemplateVariableMismatchException();
            }
        }

        if (foundKeys.size() != fieldPlaceholders.size()) {
            throw new TemplateVariableMismatchException();
        }
    }

    public boolean hasSameFields(Map<String, String> fieldPlaceholders) {
        Set<String> fields = this.fieldPlaceholders.keySet();
        Set<String> otherFields = fieldPlaceholders.keySet();

        return fields.size() == otherFields.size() && fields.containsAll(otherFields);
    }

    public Set<String> getFields() {
        return fieldPlaceholders.keySet();
    }
}

File: src/main/kotlin/sharev/gathering/domain/model/IntroduceTemplateContent.kt
package sharev.gathering.domain.model

import sharev.gathering.domain.exception.GatheringException
import sharev.gathering.domain.exception.GatheringExceptionCode as GatheringCode

data class IntroduceTemplateContent(
    val text: String,
    val fieldPlaceholders: Map<String, String>,
) {
    init {
        validateContentAndFieldPlaceholders(
            text,
            fieldPlaceholders
        )
    }

    fun hasSameFields(fieldPlaceholders: Map<String, String>): Boolean {
        val fields = this.fieldPlaceholders.keys
        val otherFields = fieldPlaceholders.keys
        return fields.size == otherFields.size && fields.containsAll(otherFields)
    }

    fun getFields(): Set<String> = fieldPlaceholders.keys

    companion object {
        private val variablePattern = PatternCache.variablePattern

        private fun validateContentAndFieldPlaceholders(text: String, fieldPlaceholders: Map<String, String>) {
            val matcher = variablePattern.matcher(text)
            val foundKeys = mutableSetOf<String>()

            while (matcher.find()) {
                val key = matcher.group(1)
                foundKeys.add(key)

                if (!fieldPlaceholders.containsKey(key)) {
                    throw GatheringException(GatheringCode.WRONG_TEMPLATE)
                }
            }

            if (foundKeys.size != fieldPlaceholders.size) {
                throw GatheringException(GatheringCode.WRONG_TEMPLATE)
            }
        }
    }
}

private object PatternCache {
    val variablePattern = "\\$\\{([^}]+)}".toPattern()
}
